extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CosmosStore.java

```diff
@@ -0,0 +1,154 @@
+import org.dataportabilityproject.spi.cloud.storage.JobStore;
+import org.dataportabilityproject.spi.cloud.types.PortabilityJob;
+import org.dataportabilityproject.types.transfer.models.DataModel;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.UUID;
+
+import static org.dataportabilityproject.cloud.microsoft.cosmos.MicrosoftCloudConstants.DATA_TABLE;
+import static org.dataportabilityproject.cloud.microsoft.cosmos.MicrosoftCloudConstants.JOB_TABLE;
+
+/**
+ * A {@link JobStore} backed by Cosmos DB. This implementation uses the DataStax Cassandra driver to communicate with Cosmos DB.
+ */
+public class CosmosStore implements JobStore {
+    static final String JOB_INSERT = "INSERT INTO  " + JOB_TABLE + " (job_id, job_data) VALUES (?,?)";
+    static final String JOB_QUERY = "SELECT * FROM " + JOB_TABLE + " WHERE job_id = ?";
+    static final String JOB_DELETE = "DELETE FROM " + JOB_TABLE + " WHERE job_id = ?";
+    static final String JOB_UPDATE = "UPDATE " + JOB_TABLE + "SET job_data = ? WHERE job_id = ?";
+
+    static final String DATA_INSERT = "INSERT INTO  " + DATA_TABLE + " (data_id, data_model) VALUES (?,?)";
+    static final String DATA_QUERY = "SELECT * FROM " + DATA_TABLE + " WHERE data_id = ?";
+    static final String DATA_DELETE = "DELETE FROM " + DATA_TABLE + " WHERE data_id = ?";
+    static final String DATA_UPDATE = "UPDATE " + DATA_TABLE + "SET data_model = ? WHERE data_id = ?";
+
+    private Session session;
+    private ObjectMapper mapper;
+
+    public CosmosStore(Session session, ObjectMapper mapper) {
+        this.session = session;
+        this.mapper = mapper;
+    }
+
+    public void close() {
+        if (session != null) {
+            session.close();
+        }
+    }
+
+    @Override
+    public void create(PortabilityJob job) {
+        if (job.getId() != null) {
+            throw new IllegalStateException("Job already created: " + job.getId());
+        }
+        UUID uuid = UUID.randomUUID();
+        job.setId(uuid.toString());
+        create(uuid, job, JOB_INSERT);
+    }
+
+    @Override
+    public void update(PortabilityJob job) {
+        if (job.getId() == null) {
+            throw new IllegalStateException("Job not persisted: " + job.getId());
+        }
+        update(job.getId(), job, JOB_UPDATE);
+    }
+
+    @Override
+    public PortabilityJob find(String id) {
+        return findData(PortabilityJob.class, id, JOB_QUERY, "job_data");
+    }
+
+    @Override
+    public void remove(String id) {
+        remove(id, JOB_DELETE);
+    }
+
+    @Override
+    public <T extends DataModel> void create(String jobId, T model) {
+        create(UUID.fromString(jobId), model, DATA_INSERT);
+    }
+
+    @Override
+    public <T extends DataModel> void update(String jobId, T model) {
+        update(jobId, model, DATA_UPDATE);
+    }
+
+    @Override
+    public <T extends DataModel> T findData(Class<T> type, String id) {
```

Comment:
As I'm porting the Google cloud solution over to JobStore, I'm implementing "PortabilityJob find(String jobId)" directly and don't seem to need the generic findData.

Since we're only storing PortabilityJobs in the store as of now, can we get rid of the generic findData?

Reply:
same thing goes for the other generic type methods - create and update. Can we just have the PortabilityJob versions?

Reply:
I was planning that as a generic way for importers/exporters to store data but maybe the mechanism should just be ContinuationData? What do you think about having JobStore persist continuation data, or is there a better way to handle that?

On a related note, there is a streaming version of the persistence method on JobStore as well. My thought there was to provide a mechanism to directly stream large amounts of data from the exporter to the store and from the store to the importer based on a key that can be provided by continuation data. 

Chuy and I has a discussion about this a few weeks back and we talked about the fact that we need some form of streaming storage since the exporters and importers may be running in difference processes/containers and hence the input and output streams cannot be directly chained.

Reply:
I should also add that if the JobStore persists all job-related data, it could be that multiple backing services are used. For example in the Microsoft implementation, I was planning on using Cosmos DB for the Job data and the separate Blob Storage service for large data. The Microsoft implementation of JobStore would composite those two services but clients won't need to know about that.

Reply:
ah ok, I haven't been following the ContinuationData discussion so I'm not sure, but would it be OK to leave as a PortabilityJob store at first so it's easier to migrate the existing impls over, then we can add ContinuationData / streaming later as necessary?

This is what I had in mind: https://github.com/google/data-portability/pull/112

Reply:
Sounds good to me!